doc/adr/0004-nats-headers.md

```diff
@@ -18,7 +18,16 @@ The salient points of the HTTP header specification are:
 - The specification may allow any number of strange things like comments/tokens etc.
 - The keys can repeat.
 
-The reference NATS Go client uses the `http.Header` infrastructure for serializing NATS headers. Other language clients may not be able to reuse header serialization infrastructure outside of the context of an HTTP request. Furthermore, the implementation of the native header library may diverge from the Go `http.Header` while still being standard. 
+More specifically from [rfc822](https://www.ietf.org/rfc/rfc822.txt) Section 3.1.2:
+> Once a field has been unfolded, it may be viewed as being composed of a field-name followed by a colon (":"), followed by a field-body, and  terminated  by  a  carriage-return/line-feed.
+> The  field-name must be composed of printable ASCII characters (i.e., characters that  have  values  between  33.  and  126., decimal, except colon).  The field-body may be composed of any ASCII characters, except CR or LF.  (While CR and/or LF may be present  in the actual text, they are removed by the action of unfolding the field.)
+
+The reference NATS Go client uses the `http.Header` infrastructure for serializing NATS headers. Other language clients may not be able to reuse header serialization infrastructure outside of the context of an HTTP request. Furthermore, the implementation of the native header library may diverge from the Go `http.Header` while still being standard.
+
+For example, the Go `http.Header` library, performs the following formatting which the while not part of the spec, as header field names are case insensitive, are a requirement in NATS header fields:
+- First character in a header field is capitalized if in the range of [a-z]
+- First characters following a dash (`-`) are capitalized if in the range of [a-z]
+_If a client doesn't serialize field names as above, it is possible that the server may ignore them._
```

Comment:
So what do we do if a user sends in `foo-bar: baz` header? Do we "fix" it to our broken ways or leave as is with potential ramifications or error on invalid header shape?

Reply:
header must be Foo-Bar for the server if we are not honoring case insensitive.

Reply:
We canot change what users give us, we cannot require all headers to comply.

Reply:
> performs the following formatting which the while not part of the spec

Typo (which the while). Precision, it is not the serialization that does the canonicalization of the keys. The canonicalization occurs if you use m.Header.Set() or m.Header.Add(). So it happens in those APIs before setting the keys in the map. As discussed on Slack yesterday, if the user does `m.Header["key"]=[]string{"val"}` then you will see that Go does not change the case.
The conclusion yesterday was that if users deal with headers that the server cares about, the user will be expected to set the key with the case that the server expects, say `Msg-Id`. Otherwise, user is free to use any case it wants. So, it would mean that libraries are actually not required to implement canonical keys.

Reply:
Here are headers from Jaeger:

```
uber-trace-id: {trace-id}:{span-id}:{parent-span-id}:{flags}
```

And from Trace Context

```
traceparent: 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
tracestate: rojo=00f067aa0ba902b7,congo=t61rcWkgMzE
```

Trace Context is a new w3 standard for tracing:

```
Vendors MUST expect the header name in any case (upper, lower, mixed), and SHOULD send the header name in lowercase.
```

now this implies that us fiddling the headers case is acceptible - as it should be since headers are case insensitive.  As a general though I think it's really user unfriendly when they give in x-y-z and we map that to X-Y-Z arbitrarily just because its a bit faster.

Reply:
Worse: The Go client lift the headers using ReadMIMEHeader(), which means it is going to convert keys to canonical form. So in theory, if you add a key to the received message and send that message back, you will have changed possibly existing keys that are lower case to canonical. But again, http.Header keys are supposed to be case insensitive. So it should not matter which case we use. This is why I think a safe approach is to use Go's way of decoding keys with conversion to canonical and only offer canonical form of setting keys. This way, the server will always find the appropriate key (when doing low level header scan that it already does for the JetStream de-dup of Msg-Id).

Reply:
>those APIs before setting the keys in the map. As discussed on Slack yesterday, if the user does `m.Header["key"]=[]string{"val"}` then you will see that Go does not change the case.

this may be a detail on the implementation on how they load the keys. If the loading of the headers uses set/add/etc it will deal with the keys.

The keys from the client perspective are case-insensitive. `get('foo')` is same as `get('Foo')`, and that is the API contract.

Reply:
By far the prefered best outcome is if the server handled headers as per specification, then none of this matters.

Reply:
Or more easy, we make the spec borrowing the ideas, but not the language implementation for that. Then we can just say x/y. The Mime issues introduced by the type is a problem, HTTP headers are not MIME, but looks like go uses them as if they were - that creates all sorts of weird things spaces allowed in keys, and line folding.

Reply:
From a product perspective I think we need to be flexible in terms of what clients can exchange.  Mirroring raw HTTP header support will get us close which will help us in the future.  Like HTTP, clients that need to include restricted characters would need to encode.

@derekcollison mentioned in a slack conversation that it's reasonable for the server to expect a particular format and casing (i.e. Canoncial form) of a header key - I agree.  Admittedly it could yield confusion, e.g. mistakenly using "msg-id" vs "Msg-Id", but I believe those are corner cases that are few and far between.

Expanding on this conversation, are we going to support quoted strings to allow `,` or `:` as part of a value?

Reply:
The value can be anything that it wants - as per the spec, the only thing the value cannot have is `\r` and `\n`. In Go, it seems these get stripped and replaced by 2 spaces each, @kozlovic  can clarify as he was looking at that.

Reply:
IIRC, the spec actually says the server should reject that case.  It sound like per [RFC822](https://www.ietf.org/rfc/rfc822.txt), quoted strings are supported, correct?

Reply:
@ColinSullivan1 - the main issue is really whether we require the Canonical Form - from a client's perspective there's no difference, but the serialized header name must match what the server is expecting if the server is not lenient on the parsing. 

Also note the obvious from the example above, that if messages are forwarded, because system headers are not prefixed, it could easily be clobbered by an application. I think on our side, we should be prefixing - whether others do or don't is an application issue.